lib/io/stdin.fz

```diff
@@ -45,53 +45,96 @@ private stdin(
+  # note: the generator function is called only once for hasNext/next
+  #
+  private stream_from(T type, generator () -> T, allowed_next T -> bool) stream T is
 
-    val option<outcome<u8>> := nil
+    val := mut (option T nil)
 
     read =>
-      if val.isNil
-        set val := stdin.nextByte
+      if val.get.isNil
+        val <- generator()
 
     reset =>
-      set val := nil
+      val <- nil
 
-    ref: stream<u8>
+    ref: stream T
       redef hasNext bool is
         read
-        match val
+        match val.get
           nil => fuzion.std.panic "this should not be possible"
-          v outcome<u8> =>
-            match v
-              error => false
-              v u8 => true
-      redef next u8 is
+          v T => allowed_next v
+      redef next T is
         read
-        match val
+        match val.get
           nil => fuzion.std.panic "this should not be possible"
-          v outcome<u8> =>
-            match v
-              error => fuzion.std.panic "no next value in stream"
-              v u8 =>
-                reset
-                v
+          v T => reset; v
 
 
   # read n codepoints from stdin
-  read(n i32) string is
+  # may return less than n bytes if end of file is reached
+  # or error if stdin can not be read right now
+  read(n i32) outcome string
+  pre n >= 0
+  is
+    any_error := mut (choice error nil) nil
     tmp : string is
-      redef utf8 Sequence<u8> is byte_stream
-                                                            # asArray, since we don't want this to be lazy
-    strings.fromCodepoints (tmp.asCodepoints.asStream.take n).asArray
+      is_next (choice u8 end_of_file error) -> bool := ((next) ->
+        match next
+          u8 u8 => true
+          end_of_file end_of_file => false
+          error error => any_error <- error; false)
+      redef utf8 Sequence u8 is
+        (stream_from (() -> nextByte) is_next)
+          .map (v ->
+            match v
+              u8 u8 => u8
+              * => panic "not defined"; u8 0)
+
+    arr := (tmp.asCodepoints.asStream.take n)
+    #  asArray, since we don't want this to be lazy
+      .asArray
+
+    match any_error.get
+      error error => error
+      * =>
+        strings.fromCodepoints arr
+
+
+  # read from stdin until end of line or end of file
+  # returns an error if stdin is already depleted = end of file
+  # NYI read_line should probably filter carriage returns
+  read_line outcome string is
+    any_error := mut (choice error nil) nil
+    got_end_of_file := mut false
+
+    is_next (choice u8 end_of_file error) -> bool := ((next) ->
+      match next
+        u8 u8 => u8 != line_feed
+        end_of_file end_of_file => got_end_of_file <- true; false
+        error error => any_error <- error; false)
+
+    bytes := (stream_from (() -> nextByte) is_next)
+      .map (v ->
+        match v
+          u8 u8 => u8
+          * => panic "not defined"; u8 0)
```

Comment:
is the `u8 0` required here? panic returns`void` which should be assignable to any other type.
`fz` front end should maybe even report an error since the code `u8 0` is unreachable, can you create an issue for this?